src/evaluators/JSXElement.js

```diff
@@ -0,0 +1,331 @@
+      default:
+        invariant(false, `Unknown JSX attribute type:: ${astAttribute.type}`);
+    }
+  }
+  return {
+    attributes,
+    children,
+  };
+}
+
+function getReactElementSymbol(realm: Realm) {
+  if (reactElementSymbol !== undefined) {
+    return reactElementSymbol;
+  }
+  let SymbolFor = realm.intrinsics.Symbol.properties.get("for");
+  if (SymbolFor !== undefined) {
+    let SymbolForDescriptor = SymbolFor.descriptor;
+
+    if (SymbolForDescriptor !== undefined) {
+      let SymbolForValue = SymbolForDescriptor.value;
+      if (SymbolForValue !== undefined && typeof SymbolForValue.$Call === "function") {
+        reactElementSymbol = SymbolForValue.$Call(realm.intrinsics.Symbol, [
+          new StringValue(realm, reactElementSymbolKey),
+        ]);
+      }
+    }
+  }
+  return reactElementSymbol;
+}
+
+function createReactProps(
+  realm: Realm,
+  type: Value,
+  attributes: Map<string, Value>,
+  children,
+  env: LexicalEnvironment
+): ObjectValue {
+  let obj = ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+  for (let [key, value] of attributes) {
+    if (typeof key === "string") {
+      if (RESERVED_PROPS.hasOwnProperty(key)) {
+        continue;
+      }
+      CreateDataPropertyOrThrow(realm, obj, key, value);
+    }
+  }
+  if (children !== null) {
+    CreateDataPropertyOrThrow(realm, obj, "children", children);
+  }
+  return obj;
+}
+
+function createReactElement(realm: Realm, type: Value, key: Value, ref: Value, props: ObjectValue) {
+  let obj = ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+  CreateDataPropertyOrThrow(realm, obj, "$$typeof", getReactElementSymbol(realm));
+  CreateDataPropertyOrThrow(realm, obj, "type", type);
+  CreateDataPropertyOrThrow(realm, obj, "key", key);
+  CreateDataPropertyOrThrow(realm, obj, "ref", ref);
+  CreateDataPropertyOrThrow(realm, obj, "props", props);
+  CreateDataPropertyOrThrow(realm, obj, "_owner", realm.intrinsics.null);
+  return obj;
+}
+
+export default function(
+  ast: BabelNodeJSXElement,
+  strictCode: boolean,
+  env: LexicalEnvironment,
+  realm: Realm
+): Value | Reference {
+  invariant(realm.react.enabled, "JSXElements can only be evaluated with the reactEnabled option");
+  let openingElement = ast.openingElement;
+  let type = evaluateJSXIdentifier(openingElement.name, strictCode, env, realm);
+  let { attributes, children } = evaluateJSXAttributes(
+    openingElement.name,
+    openingElement.attributes,
+    ast.children,
+    strictCode,
+    env,
+    realm
+  );
+  let key = attributes.get("key") || realm.intrinsics.null;
+  let ref = attributes.get("ref") || realm.intrinsics.null;
+
+  if (key === realm.intrinsics.undefined) {
+    key = realm.intrinsics.null;
+  }
+  if (ref === realm.intrinsics.undefined) {
+    ref = realm.intrinsics.null;
+  }
+
+  if (key !== realm.intrinsics.null && key instanceof ConcreteValue) {
+    key = new StringValue(realm, ToString(realm, key));
```

Comment:
If `key` is an abstract value, is it okay to be not a string?

Reply:
I believe it should be okay. @sebmarkbage can confirm.

Reply:
I don't understand the reasoning behind this. Even if it's okay, a comment in the code seems appropriate.

Reply:
This will turn the key into `null` for an abstract value. I don't think that's what we want in that case.

The actual code for this is `'' + key`. That case is already handled by Prepack.

You could do the same by calling `computeBinary(realm, '+', new StringValue(realm, ''), key)`. https://github.com/facebook/prepack/blob/336236d85ba90686403c12b4040d23150ac8f953/src/evaluators/BinaryExpression.js#L148-L156

For an abstract value it'll yield a new abstract value that is the computation `'' + key`.

Reply:
Made this change, thanks for the help.